src/main/java/com/artipie/maven/aether/AetherRepository.java

```
@@ -0,0 +1,173 @@
+        /**
+         * ServiceLocator instance.
+         */
+        private final ServiceLocator locator;
+
+        /**
+         * Ongoing session.
+         */
+        private final RepositorySystemSession session;
+
+        /**
+         * All args constructor.
+         * @param locator ServiceLocator instance
+         * @param session Ongoing session
+         */
+        private Deployer(final ServiceLocator locator, final RepositorySystemSession session) {
+            this.locator = locator;
+            this.session = session;
+        }
+
+        /**
+         * Transfers given content to staging file and deploys it.
+         * @param path Artifact URI path segment
+         * @param content Artifact binary
+         * @return DeployResult
+         * @throws Exception Deployment failed
+         */
+        private DeployResult deploy(final String path, final InputStream content) throws Exception {
+            final var repositories = this.locator.getService(RepositorySystem.class);
+            if (repositories == null) {
+                throw new IllegalStateException("RepositorySystem cannot be null");
+            }
+            DeployResult result = null;
+            try (var staging = AetherRepository.this.parent.resolve(path)) {
+                try (var file = Files.newOutputStream(staging.unwrap())) {
+                    content.transferTo(file);
+                }
+                result = repositories.deploy(
+                    this.session,
+                    new DeployRequest()
+                        .addArtifact(
+                            new DefaultArtifact(
+                                new FileCoordinates(path).coords()
+                            ).setFile(
+                                staging.unwrap().toFile()
+                            )
+                        )
+                );
+            } catch (final FileCleanupException ex) {
+                LOG.warn("on AutoCloseablePath cleanup", ex);
```

@Nimmaster where is this exception thrown from exactly? Is it ok to swallow it?

---

We had this discussion. It's just a temporary file, its deletion means the work has already done, we should not fail just because of it.
It fails on creating and reading the file as usual.

---

@Nimmaster And what happens if the File cannot be deleted? What happens if this "cleanup" fails? Will the files remain on disk? What about the locking?

If we happen to recreate the same file later, what kind of IOExceptions will we get because the file is already present and should have been deleted long ago?

---

We can delete it later.
There could be a logic to generate a randomized name or to create file in another directory, or to clean the whole temporary directory on startup.